Improve for each

gruenich opened the pull request.

gruenich commented:
If you are interested, there are a couple of `for` loops with classical iterators that could be converted to range-based for loops. I can provide according changes.

shundhammer commented on src/PkgCommitPage.cc:
```diff
@@ -363,7 +363,8 @@ void PkgCommitPage::initProgressData()
 
-    foreach ( PkgTask * task, pkgTasks()->todo() )
+    const PkgTaskList tasks = pkgTasks()->todo();
+    for ( const PkgTask * task : tasks )
```
Please add a blank line before the `for()`.

shundhammer replied:
On second thought, please let's leave that as the one-liner that it was; creating another variable that the next line refers to (and then is never used again) adds to the mental burden of reading the code.

And as I wrote elsewhere here, the compiler will generate the exact same code; "common subexpression" optimization is the simplest one there is, one that every compiler always does. For sure the functions will not be called several times; alone for the fact that there might be unexpected side effects when doing that.

gruenich replied:
Added blank line. Keeping the local variable for correctness, not performance.

shundhammer commented on src/Workflow.cc:
```diff
@@ -38,7 +38,7 @@ Workflow::Workflow( const WorkflowStepList & steps )
     checkDuplicateIds();
 
-    foreach ( WorkflowStep * step, _steps )
+    for ( WorkflowStep * step : std::as_const( _steps ) )
```
Why do we need "syntactic bitter pills" (the opposite to "syntactic sugar") here? This makes the code worse to read. Is there any advantage to offset this?

Even if this might save a handful of CPU cycles (which I am not sure about), those loops never run for more than a handful (< 10) iterations. We would be trading code readability for a miniscule performance improvement, if there even is any.

shundhammer commented on src/YQPkgConflictList.cc:
```diff
@@ -76,22 +74,19 @@ YQPkgConflictList::clear()
-    ZyppProblemList::iterator it = problemList.begin();
-
-    while ( it != problemList.end() )
+    for (ZyppProblem problem : problemList )
```
Please add a blank after the opening parenthesis: `for ( ZyppProblem ...`.

shundhammer commented on src/YQPkgStatusFilterView.cc:
```diff
@@ -63,8 +63,9 @@ void YQPkgStatusFilterView::connectWidgets()
+    const auto& children = findChildren<QCheckBox *>();
 
-    foreach ( QCheckBox * checkBox, findChildren<QCheckBox *>() )
+    for ( QCheckBox * checkBox : children )
```
I deliberately chose not to do this. Why spread this over several lines?

I am pretty sure that the compiler will optimize this subexpression out of the loop anyway; there will certainly be no more than one function call.

And `auto` variables make things worse for anybody reading the code; it basically signals something along the lines of "I have no idea what type this thng returns". This can be totally avoided by leaving the function call in the `for` loop.

shundhammer commented on src/YQPkgTechnicalDetailsView.cc:
```diff
@@ -81,15 +81,13 @@ YQPkgTechnicalDetailsView::authorsListCell( ZyppPkg pkg ) const
+    const std::list<std::string> authors = pkg->authors();
 
-    while ( it != authors.end() )
+    for (const std::string& author : authors )
```
Please add a blank after the opening parenthesis and before the `&`: `for( const std::string & author`.

shundhammer commented:
Back in the days, Qt's `foreach` loop was a novel thing that could improve a lot over old code with a clumsy `for` loop with a `begin()` and an `end()` iterator. It might even have given rise to the concept of modern C++ ranged `for` loops. Both are basically equivalent; I am very sure the compiler produces the same code.

Do we have to replace the old `foreach` loops now? I don't know. We can, but I am not sure if there are compelling reasons. It's becoming more modern C++ if we do, I guess; but still, we are using Qt things all over the place anyway, from QString to all kinds of Qt container classes. We can do that change; I don't have any strong feelings one way or the other.

shundhammer commented on src/Logger.cc:
```diff
@@ -314,9 +314,9 @@ static void qt_logger( QtMsgType msgType,
-    QStringList lines = msg.split("\n");
+    const QStringList lines = msg.split( "\n" );
 
-    foreach ( QString line, msg.split( "\n" ) )
```
That was indeed duplicated code. Good find!

shundhammer commented on src/MainWindow.cc:
```diff
@@ -133,7 +133,7 @@ QWidget * MainWindow::findPage( const QString & pageName )
     }
 
-    foreach ( QWidgetPointer page, _pages )
+    for ( const QWidgetPointer& page : std::as_const( _pages ) )
```
Let's not get overboard here. `QWidgetPointer` is typedef`ed to a `QPointer<QWidget>`. It's still a pointer type. It might carry a reference count in addition to the pointer itself, but it's still a very trivial data type, not anything that costs enormous memory overhead. Those types should not be used as a const reference; that only makes the code less readable.

And again, `std::const()` adds to making the code harder to read.

gruenich replied:
Without the reference, it will create a copy of a `QWidgetPointer` in every loop iteration.

shundhammer commented on src/YQPkgDescriptionView.cc:
```diff
@@ -136,7 +136,7 @@ QString YQPkgDescriptionView::simpleHtmlParagraphs( QString text )
     QString html_text = "<p>";
 
-    QStringList lines = text.trimmed().split( '\n', Qt::KeepEmptyParts );
+    const QStringList lines = text.trimmed().split( '\n', Qt::KeepEmptyParts );
```
Not sure if this `const` does anything but making the code harder to read. It's not as if the body here would even attempt to modify this variable.

Const references have a use in function parameters: It's not only a promise, but an assurance that the variable will not be modified in the function.

But inside a function, it just adds to the mental burden of reading the code; more information to keep in mind, to think about why this was chosen. It distracts from the code's intent.

shundhammer commented on src/YQPkgFilterTab.cc:
```diff
@@ -626,7 +626,7 @@ YQPkgFilterTab::readSettings()
     settings.beginGroup( "Pages" );
 
-    QStringList pages = settings.value( "OpenPages"   ).toStringList();
+    const QStringList pages = settings.value( "OpenPages"   ).toStringList();
```
Same as above.

shundhammer commented:
Some general things about the coding style: I try to keep it consistent and readable with const references as `const QString & text`, i.e. a blank before and after the `&`.

I know that different styles exist; there are pros and cons for most of them. It could be argued that the `&` is similar to a pointer, hence some use `const QString &text` (similar to `char *text` in old style C). But in this project we had started with the old YaST C++ coding style which was actually quite consistent, and I try to keep it consistent and clean it up whenever I find something.

As for `foreach` vs. ranged `for`: Please let's keep the punctuation consistent; like in human language and in similar places like `switch... case FOO: doSomething()`: No blank before, one blank after; like `void myfunc( int a, char * b, bool c )`, i.e. `for ( const QString line: lines )`.

Yes, I am nitpicky. I know. That's me. :smiley:

gruenich commented:
The extra variable and `std::as_const` are not about correctness. The Qt containers can detach, see [Qt documentation](https://doc.qt.io/qt-6/foreach-keyword.html) or an [insightful blog post](https://www.kdab.com/goodbye-q_foreach/). Qt's `fooreach` might be a bit more convenient, Qt prefers the plain C++11 `for` loops. For future safety, I would prefer to apply the change.

I am fine with Myrlyn's style-guide. I just sometimes forget about it, please ask for fixes just as you did.

gruenich commented:
I fixed all the white-space stuff you remarked. It is your call if you accept the reduced readability to get more standard C++ code. I think it is worth it, but it is your call.

gruenich commented:
I addressed all your comments and rebased the branch. Ready to be merged.

If you do not want to replace the `foreach` by `for` loops with additional `const` code to prevent detaching, please tell me. Then I will remove these parts from the second commit.

shundhammer commented on src/YQPkgConflictList.cc:
```diff
@@ -76,22 +74,19 @@ YQPkgConflictList::clear()
 
 void
-YQPkgConflictList::fill( ZyppProblemList problemList )
+YQPkgConflictList::fill( const ZyppProblemList problemList )
```
We might as well go all the way here and use a const &.

shundhammer commented:
Background article:

https://www.kdab.com/blog-qasconst-and-stdas_const/

shundhammer commented:
See also https://doc.qt.io/qt-6/implicit-sharing.html

shundhammer commented:
In almost all cases this is a purely theoretical problem: The containers used here are mostly tiny. We are talking about a one-digit size in most cases. So even if sometimes they _will_ detach, the performance impact is miniscule; that does in no way justify putting the burden of needing to understand the gory details of `std::as_const()` to those reading the code.

It would be a completely different matter for cases where the loop has to iterate over _thousands_ or ten thousands of elements (even hundreds would be negligible). Iterating over the zypp pool of solvables (usually packages) would be such a case. But not something like a handful of workflow steps or similar.

And if it's about containers like QList, it's only about one pointer per element; IIRC QList does not store the whole value type, just a pointer to it if the value type is more than the size of a pointer.

So, bottom line: If we can get rid of that `std::as_const()`, I am all for merging those changes.

gruenich commented:
I removed all `std::as_const` and added three more references for copies of containers we iterate over.

shundhammer reviewed (approved):
LGTM now.

Thank you for your patience!

gruenich commented:
> Thank you for your patience!

Same to you. Thanks for reviewing all my litte changes and uncovering my mistakes!
